lib/model/internal_link.dart

```diff
@@ -0,0 +1,282 @@
+}
+
+/// A [Narrow] from a given URL, on `store`'s realm.
+///
+/// Returns `null` if any of the operator/operand pairs are invalid.
+///
+/// Since narrow links can combine operators in ways our [Narrow] type can't
+/// represent, this can also return null for valid narrow links.
+///
+/// This can also return null for some valid narrow links that our Narrow
+/// type *could* accurately represent. We should try to understand these
+/// better, but some kinds will be rare, even unheard-of:
+///   #narrow/stream/1-announce/stream/1-announce (duplicated operator)
+Narrow? parseInternalLink(Uri url, PerAccountStore store) {
+  if (!_isInternalLink(url, store.account.realmUrl)) return null;
+
+  final (category, segments) = _getCategoryAndSegmentsFromFragment(url.fragment);
+  switch (category) {
+    case 'narrow':
+      if (segments.isEmpty || !segments.length.isEven) return null;
+      return _interpretNarrowSegments(segments, store);
+  }
+  return null;
+}
+
+/// Check if `url` is an internal link on the given `realmUrl`.
+bool _isInternalLink(Uri url, Uri realmUrl) {
+  if (url.hasScheme || url.hasAuthority || url.hasPort) {
+    // input that do not include scheme, host, and port
+    // are treated as on realm
+    try {
+      if (url.origin != realmUrl.origin) return false;
+    } on StateError {
+      return false;
+    }
+  }
+  return (url.hasEmptyPath || url.path == '/')
+    && !url.hasQuery
+    && url.hasFragment;
+}
+
+/// Split `fragment` of arbitrary segments and handle trailing slashes
+(String, List<String>) _getCategoryAndSegmentsFromFragment(String fragment) {
+  final [category, ...segments] = fragment.split('/');
+  if (segments.length > 1 && segments.last == '') segments.removeLast();
+  return (category, segments);
+}
+
+Narrow? _interpretNarrowSegments(List<String> segments, PerAccountStore store) {
+  assert(segments.isNotEmpty);
+  assert(segments.length.isEven);
+
+  ApiNarrowStream? streamElement;
+  ApiNarrowTopic? topicElement;
+  ApiNarrowDm? dmElement;
+
+  for (var i = 0; i < segments.length; i += 2) {
+    final (operator, negated) = _parseOperator(segments[i]);
+    // TODO(#252): negated operators are not currently supported
+    // These should eventually return a SearchNarrow
+    if (negated == true) return null;
+    final operand = segments[i + 1];
+    switch (operator) {
+      case _NarrowOperator.stream:
+        if (streamElement != null) return null;
+        final streamId = _parseStreamOperand(operand, store);
+        if (streamId == null) return null;
+        streamElement = ApiNarrowStream(streamId, negated: negated);
+
+      case _NarrowOperator.topic:
+      case _NarrowOperator.subject:
+        if (topicElement != null) return null;
+        final String topic;
+        try {
+          topic = decodeHashComponent(operand);
+        } on ArgumentError {
+          return null;
+        } on FormatException {
+          return null;
+        }
```

Comment:
This method relies on `Uri.decodeComponent` which can be easily made to throw when you pass in the wrong percentage encoding values:
```
  final cases = [
    "%1",
    "%FF",
  ];
  for (final input in cases) {
    try {
      Uri.decodeComponent(input);
    } catch (e) {
      print("$input: ${e.toString()}");
    }
  }
```

Reply:
Cool, thanks for the examples. I guess that's the same behavior as JS's (or the Web platform's?) `decodeURIComponent` — from my browser console:
```
> decodeURIComponent('%1');
VM99:1 Uncaught URIError: URI malformed
    at decodeURIComponent (<anonymous>)
    at <anonymous>:1:1
> decodeURIComponent('%FF');
VM106:1 Uncaught URIError: URI malformed
    at decodeURIComponent (<anonymous>)
    at <anonymous>:1:1
```

Kind of messy that `Uri.decodeComponent` throws two different exception types there, but so it goes.

Anyway, should be encapsulated inside our `decodeHashComponent` helper rather than having each caller have to deal with it.

Reply:
Oh and to complete your examples — here's the output:
```
%1: Invalid argument(s): Truncated URI
%FF: FormatException: Invalid UTF-8 byte (at offset 0)
```